mailet/standard/src/main/java/org/apache/james/transport/matchers/AtMost.java

```diff
@@ -0,0 +1,73 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport.matchers;
+
+import java.util.Collection;
+import java.util.Optional;
+
+import javax.mail.MessagingException;
+
+import org.apache.james.core.MailAddress;
+import org.apache.mailet.Attribute;
+import org.apache.mailet.AttributeName;
+import org.apache.mailet.AttributeUtils;
+import org.apache.mailet.AttributeValue;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.GenericMatcher;
+import org.apache.mailet.base.MailetUtil;
+
+import com.google.common.collect.ImmutableList;
+
+/**
+ * Checks that a mail did at most X tries on a specific operation.
+ *
+ * If no retries have been performed previously, it sets up an attribute `AT_MOST_TRIES`
+ * in the mail that will be incremented everything the check succeeds.
+ *
+ * The check fails when the defined X limit is reached.
+ *
+ * <p>The example below will match mail with at most 3 tries on the mailet</p>
+ *
+ * <pre><code>
+ * &lt;mailet match=&quot;AtMost=3&quot; class=&quot;&lt;any-class&gt;&quot;&gt;
+ * &lt;/mailet&gt;
+ * </code></pre>
+ */
+public class AtMost extends GenericMatcher {
+    static final AttributeName AT_MOST_TRIES = AttributeName.of("AT_MOST_TRIES");
+    private Integer atMostRetries;
+
+    @Override
+    public void init() throws MessagingException {
+        this.atMostRetries = MailetUtil.getInitParameterAsStrictlyPositiveInteger(getCondition());
+    }
+
+    @Override
+    public Collection<MailAddress> match(Mail mail) throws MessagingException {
+        return AttributeUtils.getValueAndCastFromMail(mail, AT_MOST_TRIES, Integer.class)
+            .or(() -> Optional.of(0))
+            .filter(retries -> retries < atMostRetries)
```

Comment:
I'm confused are we speaking of tries or retries ?
The line before we get the value of `AT_MOST_TRIES` so it should be a tries. But the default value is 0 which seems to be a value for a retries. And then during the filter we said it's a retries.

Reply:
Good point. 

Imo we need to be speaking of "executions" (the number of time we executed this matcher).

Speaking of tries or retries, we are leaking the usage into the implementation. I think that the use case of this should be broader than counting failures.

Does this answer your question @remk ?

Reply:
Good point yes @remk ! Mixed up in the code. Was first retries but then thinking was too specific to the use we want at first, while this should be more generic, thus I switched to tries, but I forgot this constant on the way.

`executions`, `excecutionsCount`, `nbOfexecutions` might make more sense yes, thanks for the suggestion

Reply:
Yes, it makes more sense now

Reply:
Done.